src/levels.rs

```diff
@@ -1,5 +1,1068 @@
+            this_level.delete_tables(&compact_def.top)?;
+            next_level.replace_tables(&compact_def.bot, &new_tables)?;
+        } else {
+            let mut this_level = this_level.write().unwrap();
+            this_level.delete_tables(&compact_def.top)?;
+            this_level.replace_tables(&compact_def.bot, &new_tables)?;
+        }
+
+        // TODO: Add log.
+
+        Ok(())
+    }
+
+    // Picks some tables on that level and compact it to next level.
+    fn do_compact(
+        self: &Arc<Self>,
+        idx: usize,
+        mut cpt_prio: CompactionPriority,
+        pool: Arc<yatp::ThreadPool<yatp::task::callback::TaskCell>>,
+    ) -> Result<()> {
+        let level = cpt_prio.level;
+
+        assert!(level + 1 < self.opts.max_levels);
+
+        if cpt_prio.targets.base_level == 0 {
+            cpt_prio.targets = self.level_targets();
+        }
+
+        let mut compact_def;
+
+        if level == 0 {
+            let targets = cpt_prio.targets.clone();
+            let base_level = targets.base_level;
+            let next_level = self.levels[base_level].clone();
+            compact_def = CompactDef::new(
+                idx,
+                self.levels[level].clone(),
+                level,
+                next_level,
+                base_level,
+                cpt_prio,
+                targets,
+            );
+            self.fill_tables_l0(&mut compact_def)?;
+        } else {
+            let next_level = self.levels[level + 1].clone();
+            let targets = cpt_prio.targets.clone();
+            compact_def = CompactDef::new(
+                idx,
+                self.levels[level].clone(),
+                level,
+                next_level,
+                level + 1,
+                cpt_prio,
+                targets,
+            );
+            self.fill_tables(&mut compact_def)?;
+        };
+        if let Err(err) = self.run_compact_def(idx, level, &mut compact_def, pool) {
+            error!("failed on compaction {:?}", err);
+            self.cpt_status.write().unwrap().delete(&compact_def);
+        }
+
+        // TODO: Will compact_def be used now?
+
+        info!(
+            "compactor #{} on level {} success",
+            idx, compact_def.this_level_id
+        );
+        self.cpt_status.write().unwrap().delete(&compact_def);
+
+        Ok(())
+    }
+
+    fn add_l0_table(&self, table: Table) -> Result<()> {
+        if !self.opts.in_memory {
+            // Update the manifest _before_ the table becomes part of a level_handler.
+            self.manifest
+                .add_changes(vec![new_create_change(table.id(), 0, 0)])?;
+        }
+
+        while !self.levels[0].write()?.try_add_l0_table(table.clone()) {
+            let time_start = std::time::Instant::now();
+
+            while self.levels[0].read()?.num_tables() >= self.opts.num_level_zero_tables_stall {
+                std::thread::sleep(std::time::Duration::from_millis(10));
```

Comment:
In the future, we can use something like notifier or condition variable to avoid busy waiting loop.

Reply:
I added comment for it.